util/progress/progressui/display.go

```
@@ -69,53 +88,208 @@ func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch
+			return NewConsoleDisplay(c, opts...), nil
+		} else if mode == "tty" {
+			return SolveStatusDisplay{}, errors.Wrap(err, "failed to get console")
+		}
+		fallthrough
+	case PlainMode:
+		return NewPlainDisplay(out, opts...), nil
+	case RawJSONMode:
+		return NewRawJSONDisplay(out, opts...), nil
+	default:
+		return SolveStatusDisplay{}, errors.Errorf("invalid progress mode %s", mode)
+	}
+}
+
+type consoleDisplay struct {
+	t              *trace
+	disp           *display
+	width, height  int
+	displayLimiter *rate.Limiter
+}
+
+// NewConsoleDisplay creates a new SolveStatusDisplay that prints a TTY
+// friendly output.
+func NewConsoleDisplay(c console.Console, opts ...DisplaySolveStatusOpt) SolveStatusDisplay {
+	dsso := newDisplaySolveStatusOpts(opts...)
+	if dsso.phase == "" {
+		dsso.phase = "Building"
+	}
+	return SolveStatusDisplay{
+		disp: &consoleDisplay{
+			t:    newTrace(c, true),
+			disp: &display{c: c, phase: dsso.phase, desc: dsso.consoleDesc},
+		},
+	}
+}
+
+func (d *consoleDisplay) init(displayTimeout time.Duration) {
+	d.displayLimiter = rate.NewLimiter(rate.Every(displayTimeout), 1)
+}
+
+func (d *consoleDisplay) update(ss *client.SolveStatus) {
+	d.width, d.height = d.disp.getSize()
+	d.t.update(ss, d.width)
+	if !d.displayLimiter.Allow() {
+		// Exit early as we are not allowed to update the display.
+		return
+	}
+	d.refresh()
+}
+
+func (d *consoleDisplay) refresh() {
+	d.disp.print(d.t.displayInfo(), d.width, d.height, false)
+}
+
+func (d *consoleDisplay) done() {
+	d.width, d.height = d.disp.getSize()
+	d.disp.print(d.t.displayInfo(), d.width, d.height, true)
+	d.t.printErrorLogs(d.t.w)
+}
+
+type plainDisplay struct {
+	t              *trace
+	printer        *textMux
+	displayLimiter *rate.Limiter
+}
+
+// NewPlainDisplay creates a new SolveStatusDisplay that outputs the status
+// in a human-readable plain-text format.
+func NewPlainDisplay(w io.Writer, opts ...DisplaySolveStatusOpt) SolveStatusDisplay {
```

leave these private until someone has a use-case. `NewDisplay()` should be enough.

---

I'll remove it. We had a section of code that manually used the plain display and it created this awkward code:

```
d, _ = progressui.NewDisplay(os.Stderr, progressui.PlainMode)
```

That's fine it just felt weird discarding the error. But, the error is needed for `NewDisplay` because tty mode can fail and the progress mode can be invalid.